test/ex_webrtc/data_channel_test.exs

```diff
@@ -0,0 +1,170 @@
+    assert %DataChannel{id: 3, label: ^label2, protocol: ^protocol, ordered: false} = chan2
+    assert_receive {:ex_webrtc, ^pc1, {:data_channel_state_change, ^ref2, :open}}
+
+    label3 = "my label 3"
+    {:ok, %DataChannel{ref: ref3}} = PeerConnection.create_data_channel(pc2, label3)
+
+    refute_receive {:ex_webrtc, ^pc2, :negotiation_needed}
+
+    assert_receive {:ex_webrtc, ^pc1, {:data_channel, chan3}}
+    assert %DataChannel{id: 4, label: ^label3} = chan3
+    assert_receive {:ex_webrtc, ^pc2, {:data_channel_state_change, ^ref3, :open}}
+  end
+
+  describe "negotiating" do
+    test "with only channel added" do
+      {:ok, pc1} = PeerConnection.start_link()
+      {:ok, pc2} = PeerConnection.start_link()
+
+      label = "my label"
+      {:ok, %DataChannel{ref: ref1}} = PeerConnection.create_data_channel(pc1, label)
+
+      :ok = negotiate(pc1, pc2)
+      :ok = connect(pc1, pc2)
+
+      assert_receive {:ex_webrtc, ^pc2, {:data_channel, %DataChannel{label: ^label}}}
+      assert_receive {:ex_webrtc, ^pc1, {:data_channel_state_change, ^ref1, :open}}
+    end
+
+    test "with channel mixed with transceivers" do
+      {:ok, pc1} = PeerConnection.start_link()
+      {:ok, pc2} = PeerConnection.start_link()
+
+      {:ok, _sender} = PeerConnection.add_track(pc1, MediaStreamTrack.new(:audio))
+      label1 = "my label"
+      {:ok, %DataChannel{ref: ref1}} = PeerConnection.create_data_channel(pc1, label1)
+      {:ok, _sender} = PeerConnection.add_track(pc1, MediaStreamTrack.new(:video))
+
+      :ok = negotiate(pc1, pc2)
+      :ok = connect(pc1, pc2)
+
+      assert_receive {:ex_webrtc, ^pc2, {:data_channel, %DataChannel{label: ^label1}}}
+      assert_receive {:ex_webrtc, ^pc1, {:data_channel_state_change, ^ref1, :open}}
+
+      # add more tracks and channels and renegotiate
+      {:ok, _sender} = PeerConnection.add_track(pc1, MediaStreamTrack.new(:video))
+      :ok = negotiate(pc1, pc2)
+
+      label2 = "my label 2"
+      {:ok, %DataChannel{ref: ref2}} = PeerConnection.create_data_channel(pc2, label2)
+      assert_receive {:ex_webrtc, ^pc1, {:data_channel, %DataChannel{label: ^label2}}}
+      assert_receive {:ex_webrtc, ^pc2, {:data_channel_state_change, ^ref2, :open}}
+    end
+
+    test "with channel added only in renegotiation" do
+      {:ok, pc1} = PeerConnection.start_link()
+      {:ok, pc2} = PeerConnection.start_link()
+      {:ok, _sender} = PeerConnection.add_track(pc1, MediaStreamTrack.new(:audio))
+      {:ok, _sender} = PeerConnection.add_track(pc1, MediaStreamTrack.new(:video))
+
+      :ok = negotiate(pc1, pc2)
+      :ok = connect(pc1, pc2)
+
+      label = "my label"
+      {:ok, %DataChannel{ref: ref}} = PeerConnection.create_data_channel(pc2, label)
+
+      # NOTICE: in theory, this should be true, but as of now, we perform the SCTP handshake
+      # even if the SDP did not contain any data lines
+      # refute_receive {:ex_webrtc, ^pc1, {:data_channel, _}}
```

Comment:
Does it mean that we can setup datachannel without its prior negotiation? If yes, I assume this is not in the spec?

Reply:
fixed